improves docker image consistency, adds devcontainer, improves lab1

jonas-kaufmann opened the pull request:
Requires https://github.com/simbricks/simbricks/pull/3.

jonas-kaufmann committed:
fix kvm acceleration for docker in .sh scripts

jonas-kaufmann committed:
fix inconsistencies in docker image name

jonas-kaufmann committed:
Dockerfile: enable running experiments as non-root

jonas-kaufmann committed:
add vscode devcontainer

jonas-kaufmann committed:
Dockerfile: Removes lines for adding simbricks to PYTHONPATH using `ipython_config.py`

Simbricks was added to PYTHONPATH in Docker image simbricks-min using Docker's `ENV` in commit https://github.com/simbricks/simbricks/commit/fcd615a985b121d851af6bcdfbfa420e5fb5a864.
The advantage is that syntax highlighting now properly works for Jupyter cells in the VSCode devcontainer.

jonas-kaufmann committed:
lab1: add parameter names to function calls, fix typos

jonas-kaufmann committed:
reformat code in lab1

jonas-kaufmann committed:
lab1: typo and missing space

FreakyPenguin commented on docker/Dockerfile.jupyter:
```diff
@@ -7,11 +7,9 @@ RUN apt-get update \
- && echo "c.InteractiveShellApp.exec_lines = [ \
-	'import sys; sys.path.append(\"/simbricks/experiments\")' ]" \
-	>>`ipython locate`/profile_default/ipython_config.py
+	jupyterlab
+# Add simbricks experiment modules to Python path.
+ENV PYTHONPATH=/simbricks/experiments
```
Does this work both in vscode and when running the container standalone?

jonas-kaufmann replied:
Yes, it does.

FreakyPenguin replied:
Actually I'm wondering if this should not just be set in `docker/Dockerfile` and  `docker/Dockerfile.min`... As this is not really a jupyter issue. Want to put that into a PR on the simbricks repo?

jonas-kaufmann replied:
Makes sense. Maybe I can go even further and rename run.py to a proper executable name and add it to PATH. This way, it can be used  to run simbricks experiments from any directory and abstracts away simbrick's installation directory.

FreakyPenguin replied:
Indeed, that would have been my next suggestion. But as a first step I would suggest adding a "wrapper" shell script "simbricks-run" (maybe put in in the docker directory, and then in the dockerfile copy it to /usr/bin). Longer term we probably do want to just rename run.py, but I think right now we may have a few random scripts etc. laying around that invoke run.py so I'd prefer to not break these until we get an overview etc.

FreakyPenguin commented on lab1.ipynb:
```diff
@@ -2648,15 +2698,22 @@
+    "print(\n",
+    "    'Throughput test 1: %.2f Gbps\\nSimulation time: %.2fs\\n' % (\n",
+    "        parse_iperf_throughput(output,\n",
+    "                               client), output.end_time - output.start_time\n",
```
Just curious, is this what the code style explicitly suggests? Or would it be consistent with the style and more readable to break after `),` and indent to the same level as `parse` in the line above?

jonas-kaufmann replied:
The interesting section from the style guide is this one: https://google.github.io/styleguide/pyguide.html#34-indentation
Basically, we can always use either hanging indent or indent with vertical alignment. Here, both were mixed, which saves some lines but I agree, readability using your suggestion is better. I am going to look into yapf's options again and then change the formatting here.

jonas-kaufmann replied:
Quickly compared to pep8 and it seems to suggest the same.

FreakyPenguin commented on lab1.ipynb:
```diff
@@ -28,7 +28,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 35,
```
Hm could you look into into if there are any best pratices for ipython notebooks in git repos? In the current form this feels closer to having an object file stored in the repo... (not a problem with this commit though, a more general question)

jonas-kaufmann replied:
Yes, I have put that onto my TODO list.

FreakyPenguin commented:
But other than the env comment I think this is good to merge. Do you want to do a PR to the main repo first, or should we just merge this first and then fix it up?

jonas-kaufmann commented:
I'd prefer to implement it properly in the first place.

jonas-kaufmann committed:
Replaces absolute path to  `run.py` with `simbricks-run`.

`simbricks-run` was introduced by commit https://github.com/simbricks/simbricks/commit/61ba39e8be9e6cc9205562cbacb63439bf7ecd53

jonas-kaufmann committed:
makes string quotes consistent

jonas-kaufmann commented:
PR is now ready to be merged.
